assembly/src/parsers/io_ops.rs

```diff
@@ -255,70 +270,104 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+        parse_hex_param(token, 1, param_str)?
     } else {
-        if param_str.len() % HEX_CHUNK_SIZE != 0 {
-            // hex string doesn't contain a valid number of bytes
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        false
+        // If we find invalid ParamType, we have internal parsing error and should panic
+        panic!("Invalid parameter type")
     };
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(get_valid_felt(token, 1, value)?)))
+    } else {
+        Err(ParsingError::invalid_param(token, 1))
+    }
+}
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+fn parse_push_many(
+    token: &Token,
+    param_type: ParamType,
+    param_str: &str,
+) -> Result<Node, ParsingError> {
+    let values = if param_type == ParamType::Hex {
+        parse_hex_params(token, param_str)?
+    } else if param_type == ParamType::HexAndDecimal {
+        parse_constants(token)?
     } else {
-        // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
-        (0..param_str.len())
-            .step_by(HEX_CHUNK_SIZE)
-            .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
-            .collect()
+        // If we find invalid ParamType, we have internal parsing error and should panic
+        panic!("Invalid parameter type")
+    };
+    let values_max = values.iter().max();
+    match values_max {
+        Some(value) if *value <= u8::MAX as u64 => Ok(Instruction(PushU8List(
+            values.iter().map(|v| *v as u8).collect(),
+        ))),
+        Some(value) if *value <= u16::MAX as u64 => Ok(Instruction(PushU16List(
+            values.iter().map(|v| *v as u16).collect(),
+        ))),
+        Some(value) if *value <= u32::MAX as u64 => Ok(Instruction(PushU32List(
+            values.iter().map(|v| *v as u32).collect(),
+        ))),
+        Some(value) if *value <= Felt::MODULUS => {
+            let mut values_felt = Vec::new();
+            for (i, v) in values.iter().enumerate() {
+                values_felt.push(get_valid_felt(token, i + 1, *v)?)
+            }
+            if values.len() == 4 {
+                Ok(Instruction(PushWord(
+                    values_felt.try_into().expect("Invalid constatnts length"),
+                )))
+            } else {
+                Ok(Instruction(PushFeltList(values_felt)))
+            }
+        }
+        _ => Err(ParsingError::invalid_param(token, 1)),
```

Comment:
A few things here:

1. The equality should be `<` (rather than `<=`) as value which is equal to field modulus is not a valid field element.
2. Similar to my other comment, there is no need to use `get_valid_felt()` here.
3. I think we already have a `WORD_LEN` (or something like this) constant somewhere. Would be better to use it instead of 4.
4. Once this function is changed so that it doesn't need to handle parsing, we could replace the last error check with `unreachable!()`.

Reply:
One question about 4th point: since we mark the case when `value > Felt::MODULUS` as `unreachable!()`, we need to check that a `value` is a valid Felt before it. What is better -- check it in the parsing functions or, maybe, join two last cases and create Felt values with `get_valid_felt()` function? In the second approach we can get rid of one match branch, left unchanged functions that parses plain `u64` values and provide an informative error message without copying it to the new place.

Reply:
I would either check it during parsing (and then `unreachable!()` here is fine). Or if it is easier, I would return an error here instead of `unreachable!()`. But I would customize the error to make it more informative than just `invalid_param()` error. I would probably not use `get_valid_felt()` here.